rotkehlchen/chain/ethereum/modules/eth2/eth2.py

```diff
@@ -320,6 +327,67 @@ def _query_services_for_validator_daily_stats(
             if len(new_stats) != 0:
                 dbeth2.add_validator_daily_stats(stats=new_stats)
 
+    def query_services_for_validator_withdrawals(
+            self,
+            to_ts: Timestamp,
+    ) -> None:
+        """Goes through all saved validators, sees which need to query withdrawals
+        queries them and saves them in the DB.
+
+        May raise:
+        - RemoteError due to problems querying beaconcha.in API
+        """
+        dbeth2 = DBEth2(self.database)
+        dbevents = DBHistoryEvents(self.database)
+        result = dbeth2.get_validators_to_query_for_withdrawals(up_to_tsms=ts_sec_to_ms(to_ts))
+
+        # first get all validator indices to see which validators have exited and when
+        validator_indices = {x[0] for x in result}
+        data = self.beaconchain._query_chunked_endpoint(
+            indices_or_pubkeys=list(validator_indices),
+            module='validator',
+            endpoint=None,
+        )
+        exit_epoch = {}
+        now = ts_now()
+        for validator_entry in data:
+            if validator_entry['status'] != 'exited':
+                exit_epoch[validator_entry['validatorindex']] = now  # so that is_exit fails
+            else:
+                exit_epoch[validator_entry['validatorindex']] = validator_entry['exitepoch']
+
+        # Then fetch latest withdrawals for each
+        for validator_index, last_ts in result:
+            self._maybe_backoff_beaconchain(now=now, name='withdrawals')
+            new_data = scrape_validator_withdrawals(
+                validator_index=validator_index,
+                last_known_timestamp=last_ts,
+            )
+            now = ts_now()
+            self.validator_withdrawals_queried += 1
+            self.last_withdrawals_query_ts = now
+            for entry in new_data:
+                withdrawal = EthWithdrawalEvent(
+                    validator_index=validator_index,
+                    timestamp=ts_sec_to_ms(entry[0]),
+                    balance=Balance(amount=entry[2]),
+                    withdrawal_address=entry[1],
+                    is_exit=entry[0] >= exit_epoch[validator_index],
```

Comment:
Should we prevent against index error here? It can happen if beaconchain returns bad data for some reason I guess

Reply:
That's not possible. This is already done in the function that returns entry